peachjam/views/save_document.py

```diff
@@ -0,0 +1,153 @@
+from django.contrib.auth import get_user_model
+from django.contrib.auth.mixins import LoginRequiredMixin, PermissionRequiredMixin
+from django.http import Http404, HttpResponse, HttpResponseRedirect
+from django.shortcuts import get_object_or_404
+from django.urls import reverse, reverse_lazy
+from django.views.generic import (
+    CreateView,
+    DeleteView,
+    ListView,
+    TemplateView,
+    UpdateView,
+)
+from django.views.generic.edit import FormMixin
+
+from peachjam.forms import FolderForm, SaveDocumentForm
+from peachjam.models import CoreDocument, Folder, SavedDocument, pj_settings
+
+User = get_user_model()
+
+
+class AllowSavedDocumentMixin:
+    def dispatch(self, *args, **kwargs):
+        if not pj_settings().allow_save_documents:
+
+            raise Http404("Saving documents is not allowed.")
+        return super().dispatch(*args, **kwargs)
+
+
+class BaseFolderView(
+    AllowSavedDocumentMixin, LoginRequiredMixin, PermissionRequiredMixin
+):
+    model = Folder
+
+    def get_template_names(self):
+        if self.request.htmx:
+            return ["peachjam/_folders_list.html"]
+        return ["peachjam/folders_list.html"]
+
+    def get_queryset(self):
+        return self.request.user.folders.all()
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["ungrouped_saved_documents"] = self.request.user.saved_documents.filter(
+            folder__isnull=True
+        )
+        context["folders"] = self.get_queryset()
+        return context
+
+
+class FolderListView(BaseFolderView, ListView):
+    permission_required = "peachjam.view_folder"
+
+
+class BaseFolderFormView(BaseFolderView, FormMixin):
+    success_url = reverse_lazy("folder_list")
+    form_class = FolderForm
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs["request"] = self.request
+        return kwargs
+
+
+class FolderCreateView(BaseFolderFormView, CreateView):
+    permission_required = "peachjam.add_folder"
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        instance = Folder()
+        instance.user = self.request.user
+        kwargs["instance"] = instance
+        return kwargs
+
+
+class FolderUpdateView(BaseFolderFormView, UpdateView):
+    permission_required = "peachjam.change_folder"
+
+
+class FolderDeleteView(BaseFolderFormView, DeleteView):
```

Comment:
This view's form class shouldn't be the same as Create or Update -- deletion doesn't need to validate the details of the object being deleted. So it should extend `BaseFolderView` instead.

Reply:
changed